### ISSUE TITLE
Checkpoint and restore pyodide instance?

### ISSUE THREAD

**Issue by gnestor** [iodide-project/pyodide#380]
I'm doing some experimenting with checkpointing a WebAssembly instance to a file and restoring it. 

I have a very basic example working: https://observablehq.com/d/fcca050774d22728

I'm trying to do the same with a pyodide instance. So far, I'm getting the pyodide memory from `pyodide._module.wasmMemory` and I'm saving `pyodide._module.wasmMemory.buffer` to a `.wasm` file. I tried to replace [the path to `pyodide.asm.wasm`](https://github.com/iodide-project/pyodide/blob/master/src/pyodide.js#L292) with my newly saved file, but I'm hitting an instantiation error:

```
Uncaught (in promise) CompileError: AsyncCompile: Wasm decoding failed: unexpected end of stream @+1
```

Am I on the right track?

WIP notebook: https://observablehq.com/@gnestor/exporting-and-importing-a-pyodide-instance-to-from-a-wasm-fi

**Comment by mdboom** [comment]
`wasmMemory` is the memory buffer for the running wasm code.  A `wasm` file is WebAssembly bytecode.  I don't think the two are interchangeable like this.

Have you seen [wasm-persist](https://github.com/dfinity/wasm-persist)?  I think it's doing something similar to what you're trying to do.

**Comment by gnestor** [comment]
Thanks @mdboom! Yes, wasm-persist looks like the right solution.

Your distinction between wasm memory vs. wasm byte code makes total sense 👍

**Comment by gnestor** [comment]
One more question: Is there a way to get the wasm instance from the global `pyodide`?

**Comment by mdboom** [comment]
First -- I should have said before that this is something I'm really excited to see being worked on.  I've thought about something like this for a while but haven't had much time to work on it.

I don't know if there is a way to get the wasm instance, unfortunately.  I'm not opposed to adding it and it probably wouldn't be too hard with some modifications to `pyodide.js`.  As you've already seen, most of the stuff is under `pyodide._module` which is the `Module` thing that emscripten manages, so you might be able to find what you're looking for in the emscripten documentation.

**Comment by gnestor** [comment]
I'm re-opening this because wasm-persist isn't working yet. 

I posted a question on the emscripten Google Group: https://groups.google.com/forum/#!forum/emscripten-discuss

I'm glad to hear that you're excited about this too!

**Comment by vadimkantorov** [comment]
This Latex WebAssembly distribution does something similar for dumping/loading the latex process memory to speed up loading: https://github.com/kisonecat/web2js

**Comment by tanaka-chitete** [comment]
> First -- I should have said before that this is something I'm really excited to see being worked on. I've thought about something like this for a while but haven't had much time to work on it.
> 
> I don't know if there is a way to get the wasm instance, unfortunately. I'm not opposed to adding it and it probably wouldn't be too hard with some modifications to `pyodide.js`. As you've already seen, most of the stuff is under `pyodide._module` which is the `Module` thing that emscripten manages, so you might be able to find what you're looking for in the emscripten documentation.

Hi there, I am currently working on a self-contained Python notebook environment as part of my Honours project. I am trying to support notebook persistence and I would like to know how Pyodide could be modified to allow for internal state to be exposed and thus, saved to file. I tried using wasm-persists but it has too many source code files and dependencies which would make it infeasible to embed within a file.

**Comment by vadimkantorov** [comment]
In some other context, I've simply cloned the initialized Module.HEAP8 and then simply reset it to avoid module reinit. If it works in your context, it's the simplest

**Comment by hoodmane** [comment]
That won't work because the file system state and the dynamic linking state lives in JavaScript. Serializing the filesystem is not too hard. If any `.so` files have been loaded, it may require a fairly detailed understanding of the dynamic linking model to ensure that the GOT table and the state of the JS symbol tables are synced up properly.

**Comment by vadimkantorov** [comment]
Yep, my context was for a fully static linking (I explicitly chose for this path to make wasm handling simpler)

**Comment by hoodmane** [comment]
I think it's quite possible, but someone has to be willing to put in a fair amount of effort to get it working.

**Comment by tanaka-chitete** [comment]
@hoodmane If i was to try and give it a go myself, where should I start? The fact that dill isn't available in Pyodide necessitates another means to achieve saving/restoration.

**Comment by tanaka-chitete** [comment]
> Yep, my context was for a fully static linking (I explicitly chose for this path to make wasm handling simpler)

If possible, could you explain your project in more detail and how you achieved saving/restoration? Could this work be transferrable to Pyodide?

**Comment by tanaka-chitete** [comment]
> I think it's quite possible, but someone has to be willing to put in a fair amount of effort to get it working.

If i was to try and give it a go myself, where should I start? The fact that dill isn't available in Pyodide necessitates another means to achieve saving/restoration.

**Comment by sjjpo2002** [comment]
I created a [request ](https://github.com/pyodide/pyodide/issues/5909) to include dill in the built-packages.
